File: stasis-core/src/de/boxxit/stasis/serializer/MethodCallSerializer.java
package de.boxxit.stasis.serializer;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.Serializer;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import de.boxxit.stasis.MethodCall;

/**
 * User: Christian Fruth
 */
public class MethodCallSerializer extends Serializer<MethodCall>
{
	public MethodCallSerializer()
	{
		setAcceptsNull(false);
	}

	@Override
	public void write(Kryo kryo, Output output, MethodCall object)
	{
		Object[] args = object.getArgs();
		int argCount = args.length;

		output.writeString(object.getName());
		output.writeBoolean(object.isAssumeAuthenticated());
		output.writeInt(argCount, true);

		for (int i = 0; i < argCount; ++i)
		{
			kryo.writeClassAndObject(output, args[i]);
		}
	}

	@Override
	public MethodCall read(Kryo kryo, Input input, Class<MethodCall> type)
	{
		MethodCall methodCall = new MethodCall();
		kryo.reference(methodCall);

		String name = input.readString();
		boolean assumeAuthenticated = input.readBoolean();
		int argCount = input.readInt(true);
		Object[] args = new Object[argCount];

		for (int i = 0; i < argCount; ++i)
		{
			args[i] = kryo.readClassAndObject(input);
		}

		methodCall.setName(name);
		methodCall.setAssumeAuthenticated(assumeAuthenticated);
		methodCall.setArgs(args);

		return methodCall;
	}
}

File: stasis-ios/Stasis/MethodCallSerializer.m
//
//  MethodCallSerializer.m
//  Stasis
//
//  Created by Christian Fruth on 24.10.14.
//  Copyright (c) 2014 Boxx IT Solutions GmbH. All rights reserved.
//

#import "MethodCallSerializer.h"
#import "MethodCall.h"
#import "Kryo/Kryo.h"

@implementation MethodCallSerializer

- (BOOL)acceptsNull
{
	return NO;
}

- (void)write:(Kryo *)kryo value:(id)value to:(KryoOutput *)output
{
	MethodCall *methodCall = value;
	NSArray *args = methodCall.args;
	NSUInteger argCount = args.count;
	
	[output writeString:methodCall.name];
	[output writeBoolean:methodCall.assumeAuthenticated];
	[output writeInt:(SInt32)argCount optimizePositive:YES];
	
	for (NSUInteger i = 0; i < argCount; ++i)
	{
		[kryo writeClassAndObject:args[i] to:output];
	}
}

- (id)read:(Kryo *)kryo withClass:(Class)type from:(KryoInput *)input
{
	
	MethodCall *methodCall = [MethodCall new];
	[kryo reference:methodCall];
	
	NSString *name = [input readString];
	bool assumeAuthenticated = [input readBoolean];
	int argCount = [input readIntOptimizePositive:YES];
	NSMutableArray *args = [NSMutableArray arrayWithCapacity:argCount];
	
	for (int i = 0; i < argCount; ++i)
	{
		[args addObject:[kryo readClassAndObject:input]];
	}
	
	methodCall.name = name;
	methodCall.assumeAuthenticated = assumeAuthenticated;
	methodCall.args = args;
	
	return methodCall;
}

@end
